# Data Wrangling on Auralin Dataset

Auralin dataset is a fabricated clinical trial dataset for the purpose of demonstrating how to undertake a full data wrangling process systematically.

More information about the data...

- This dataset was constructed with the consultation of real doctors to ensure plausibility.
- This clinical trial data for an alternative insulin was inspired and closely mimics this real [clinical trial for a new inhaled insulin called Afrezza](http://care.diabetesjournals.org/content/38/12/2266.long).
- The data quality issues in this dataset mimic real, [common data quality issues in healthcare data](http://media.hypersites.com/clients/1446/filemanager/Articles/DocCenter_Problem_with_data.pdf). These issues impact quality of care, patient registration, and revenue.
- The patients in this dataset were created using this [fake name generator](http://www.fakenamegenerator.com/order.php) and do not include real names, addresses, phone numbers, emails, etc.

---

## Assess

- Visualize by scrolling
- Visualize by programming

In [1]:
pwd

'/Users/alejandrosanz/Downloads'

In [2]:
import os 
os.chdir('projects_on_GitHub/data_wrangling/diabetes_data_wrangling')

In [3]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

pd.set_option('display.max_rows', None)

In [4]:
patients.head(10)

patient_id assigned_sex given_name     surname                     address  \
0           1       female        Zoe     Wellish        576 Brown Bear Drive   
1           2       female     Pamela        Hill   2370 University Hill Road   
2           3         male        Jae      Debord       1493 Poling Farm Road   
3           4         male       Liêm        Phan         2335 Webster Street   
4           5         male        Tim     Neudorf        1428 Turkey Pen Lane   
5           6         male     Rafael       Costa          1140 Willis Avenue   
6           7       female       Mary       Adams            3145 Sheila Lane   
7           8       female     Xiuxiu       Chang  2687 Black Oak Hollow Road   
8           9         male      Dsvid  Gustafsson          1790 Nutter Street   
9          10       female     Sophie     Cabrera           3303 Anmoore Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   
5      Daytona Beach     Florida   32114.0  United States   
6            Burbank          NV   84728.0  United States   
7        Morgan Hill          CA   95037.0  United States   
8        Kansas City          MO   64105.0  United States   
9           New York    New York   10011.0  United States   

                                         contact   birthdate  weight  height  \
0           951-719-9170ZoeWellish@superrito.com   7/10/1976   121.7      66   
1          PamelaSHill@cuvox.de+1 (217) 569-3204    4/3/1967   118.8      66   
2               402-363-6804JaeMDebord@gustr.com   2/19/1980   177.8      71   
3     PhanBaLiem@jourrapide.com+1 (732) 636-8246   7/26/1951   220.9      70   
4                334-515-7487TimNeudorf@cuvox.de   2/18/1928   192.3      27   
5       386-334-5237RafaelCardosoCosta@gustr.com   8/31/1931   183.9      70   
6              775-533-5933MaryBAdams@einrot.com  11/19/1969   146.3      65   
7           XiuxiuChang@einrot.com1 408 778 3236   8/13/1958   158.0      60   
8        816-265-9578DavidGustafsson@armyspy.com    3/6/1937   163.9      66   
9  SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   194.7      64   

    bmi  
0  19.6  
1  19.2  
2  24.8  
3  31.7  
4  26.1  
5  26.4  
6  24.3  
7  30.9  
8  26.5  
9  33.4

In [5]:
treatments

given_name         surname    auralin    novodra  hba1c_start  \
0             veronika        jindrová  41u - 48u          -         7.63   
1               elliot      richardson          -  40u - 45u         7.56   
2             yukitaka        takenaka          -  39u - 36u         7.68   
3                 skye      gormanston  33u - 36u          -         7.97   
4               alissa          montez          -  33u - 29u         7.78   
5              jasmine           sykes          -  42u - 44u         7.56   
6               sophia          haugen  37u - 42u          -         7.65   
7                eddie          archer  31u - 38u          -         7.89   
8                saber          ménard          -  54u - 54u         8.08   
9                 asia         woźniak  30u - 36u          -         7.76   
10              joseph             day  29u - 36u          -         7.70   
11           kristiina          hyypiä          -  36u - 38u         7.87   
12             roxanne       andreyeva  29u - 38u          -         9.54   
13              gregor            bole          -  47u - 45u         7.61   
14              simone    baumgaertner  27u - 37u          -         7.74   
15                enco          žibrik  55u - 68u          -         7.78   
16             camilla        zaitseva  28u - 37u          -         7.53   
17                gina            cain          -  36u - 36u         7.88   
18          addolorata        lombardi          -  49u - 46u         7.75   
19              khalid        johnsrud          -  54u - 54u         8.35   
20                mile          stanić          -  47u - 48u         7.66   
21               tekla         walczak  29u - 39u          -         7.61   
22         brancaleone           russo  53u - 60u          -         8.61   
23            chiemela       tobeolisa          -  43u - 47u         7.59   
24                isac            berg  31u - 41u          -         9.68   
25              benoît          bonami          -  44u - 43u         9.82   
26              suhaim           rahal          -  49u - 47u         7.94   
27              mizuki           iwata          -  45u - 46u         7.70   
28             clinton          miller  42u - 51u          -         7.79   
29              eugene         mironov  42u - 49u          -         7.81   
30         szilveszter           totth  35u - 39u          -         7.70   
31                 bao             she          -  48u - 45u         9.09   
32               laura     ehrlichmann          -  43u - 40u         7.95   
33              cannan         cabrera          -  40u - 31u         7.94   
34           alexander       mathiesen  47u - 58u          -         7.96   
35               csaba            sági          -  35u - 29u         7.88   
36                 chỉ             lâm  45u - 48u          -         7.68   
37           władysław       wieczorek  24u - 37u          -         7.92   
38            kristján         ingason  44u - 55u          -         7.92   
39           krisztina          magyar          -  32u - 27u         7.52   
40                ásta     grímsdóttir          -  29u - 30u         7.62   
41              mahmud         kadyrov          -  44u - 43u         7.53   
42              marija        grubišić  37u - 43u          -         7.53   
43                jens         poulsen          -  36u - 32u         7.55   
44               sauli      koivuniemi  43u - 47u          -         7.67   
45              zdeněk           synek          -  31u - 31u         7.67   
46             mariana           souza  36u - 42u          -         7.86   
47           jad allah           gaber          -  38u - 41u         7.97   
48                 zak           kelly          -  38u - 38u         7.66   
49             jackson         addison          -  42u - 42u         7.99   
50             albinca         komavec          -  41u - 39u         7.89   
51      

In [6]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [7]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [8]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [9]:

patients.surname.value_counts()

Doe                  6
Taylor               3
Jakobsen             3
Tạ                   2
Cabrera              2
Bùi                  2
Liễu                 2
Tucker               2
Johnson              2
Correia              2
Lund                 2
Batukayev            2
Berg                 2
Souza                2
Schiavone            2
Woźniak              2
Grímsdóttir          2
Collins              2
Aranda               2
Nilsen               2
Cindrić              2
Gersten              2
Kadyrov              2
Dratchev             2
Ogochukwu            2
Hueber               2
Kowalczyk            2
Lương                2
Silva                2
Lâm                  2
Parker               2
Bazarova             1
Henzen               1
Yuan                 1
Guzman               1
Ferguson             1
Ménard               1
Maslov               1
Majewski             1
Pinto                1
Gibson               1
Sokołowska           1
Piirainen            1
Afamefuna  

In [10]:
patients.address.value_counts()

123 Main Street                  6
2476 Fulton Street               2
648 Old Dear Lane                2
2778 North Avenue                2
1854 Kyle Street                 1
4040 Linda Street                1
1434 Oak Ridge Drive             1
325 Eastland Avenue              1
4273 Deans Lane                  1
4704 Edsel Road                  1
2386 Linda Street                1
1942 Harry Place                 1
4192 Holly Street                1
1275 Goldie Lane                 1
4476 Center Street               1
3688 Adonais Way                 1
669 Archwood Avenue              1
3714 Ben Street                  1
2531 Eastland Avenue             1
1207 Garfield Road               1
602 Tator Patch Road             1
1109 Beechwood Drive             1
4643 Reeves Street               1
1373 Wilmar Farm Road            1
1904 Granville Lane              1
2931 Romano Street               1
260 Derek Drive                  1
3688 Glenview Drive              1
3113 Timber Ridge Ro

In [11]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
208    121.2
0      121.7
467    122.1
218    122.2
225    122.3
404    123.0
397    123.4
424    123.6
483    123.9
306    124.1
278    124.3
120    124.5
363    124.7
185    125.1
423    125.2
419    126.1
494    126.3
232    126.9
214    126.9
342    127.2
280    128.5
184    128.9
256    128.9
17     129.1
422    129.4
44     129.8
206    129.8
377    130.0
181    130.2
354    130.2
309    130.7
22     130.7
213    131.1
143    131.1
62     131.1
268    132.0
359    132.2
362    132.2
454    132.7
187    133.1
31     133.3
56     133.8
57     134.0
163    134.2
345    134.2
458    134.2
430    135.7

In [12]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [13]:
patients.address.duplicated().sum()

19

In [14]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [15]:
patients[patients[['given_name', 'surname']].duplicated()]

patient_id assigned_sex given_name surname          address      city  \
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state  zip_code        country                      contact birthdate  \
229    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
237    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
244    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
251    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
277    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   

     weight  height   bmi  
229   180.0      72  24.4  
237   180.0      72  24.4  
244   180.0      72  24.4  
251   180.0      72  24.4  
277   180.0      72  24.4

In [16]:
patients[patients.address.duplicated()][['given_name', 'surname']]

given_name     surname
29        Jake    Jakobsen
219         Mỹ       Quynh
229       John         Doe
230  Elisabeth     Knudsen
234    Martina   Tománková
237       John         Doe
242       John     O'Brian
244       John         Doe
249   Benjamin      Mehler
251       John         Doe
257        Jin        Kung
264   Wafiyyah      Asfour
269     Flavia  Fiorentino
277       John         Doe
278   Generosa       Cabán
282      Sandy      Taylor
286      Lewis        Webb
296        Chỉ         Lâm
502        Pat     Gersten

In [17]:
treatments.auralin.isnull().sum()

0

In [18]:
treatments.novodra.isnull().sum()

0

In [19]:
# check how many common columns in three tables
all_columns = pd.Series(list(patients) + \
                        list(treatments) + \
                        list(adverse_reactions))

all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [20]:
# # This also works!
# a = set(patients)
# b = set(treatments)
# c = set(adverse_reactions)
# inters = a.intersection(b).intersection(c)
# inters

### Quality Issues
####  `patients` table
- zip code is a float not a string
- zip code has four digits sometimes
- Tim Neudorf height is 27 instead of 72 in
- full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns)
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- multiple phone number formats
- Default John Doe data
- Multiple records for Jakebsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight


#### `treatments` table
- missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- lowercase given names and surnames
- missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (4s mistaken as 9s)
- Nulls represented as dashed (-) in auralin and novodra

#### `adverse_reactions` table
- lowercase given names and surnames



### Tidiness Issues

#### `patients` table
- contact column should be split into phone and email columns separately
 
#### `treatments` table
- three variables in two columns (should be like treatment, start_dose, end_dose)
- `given_name` and `surname` being redundant (Add `id` column to connect with `patients` table)
 
 
#### `adverse_reactions` table
- `given_name` and `surname` being redundant
- This table seems some kind of verbose (should be joined with `treatments` table)


## Clean

**General Process**
- Define
- Code
- Test

In [21]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

#### Missing Data Fixing

**`treatments`: missing records (280 instead of 350)**

___Define___

Import the cut treatments into a DataFrame and concatenate it with the original treatments DataFrame

___Code___

In [22]:
treatments_cut = pd.read_csv('treatments_cut.csv')

In [23]:
treatments_cut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [24]:
treatments_cut.shape[0] + treatments.shape[0]

350

In [25]:
treatments_clean = pd.concat([treatments_clean, treatments_cut], 
                             ignore_index=True)

___Test___

In [26]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [27]:
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

---
**`treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (4s mistaken as 9s)**

___Define___

Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`. 

___Code___

In [28]:
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - 
                                 treatments_clean.hba1c_end)

___Test___

In [29]:
treatments_clean.hba1c_change.sample(20)

318    0.40
340    0.37
201    0.30
77     0.42
80     0.54
47     0.45
334    0.37
274    0.39
91     0.43
140    0.37
143    0.35
74     0.40
317    0.33
172    0.35
301    0.46
197    0.44
44     0.30
7      0.34
55     0.37
89     0.42
Name: hba1c_change, dtype: float64

#### Tidiness Issue Fixing

**Problem 1**

**`patients` table: `Contact` column contains two variables: phone number and email**

___Define___

Extract the *phone number* and *email* variables from the *contact* column using regular expressions and pandas' `str.extract` method. Drop the *contact* column when done.

___Code___

In [30]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

___Test___

In [31]:
# Confirm contact column is gone
list(patients_clean)
# 'phone_number' in patients_clean.columns.values

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [32]:
patients_clean.phone_number.sample(25)

463         573 437 7334
187         773-607-2647
286                  NaN
405         541-967-9779
44          207 861 4587
115         707 262 6503
204         405-234-5350
135         440-385-5011
79          302-397-6035
132         570-698-4203
82          678-350-2390
460    +1 (510) 238-1151
408    +1 (989) 390-0285
262         713-556-4328
334         630-837-9473
4           334-515-7487
56     +1 (706) 497-0891
472         601-446-6829
53          617-317-5055
66          760-287-8295
497    +1 (918) 984-9171
189    +1 (916) 453-3601
76          612-925-3123
464         205-467-9279
364         818-372-7106
Name: phone_number, dtype: object

In [33]:
# Confirm that no emails start with an integer (regex didn't match for this)
# as ascii code for `0-9` is prior to that of `alphabets`, i.e. ord('1') < ord('a')
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

**Problem 2**

**`treatments` table: Three variables in two columns (treatment, start dose and end dose)**

___Define___

Melt the *auralin* and *novodra* columns to a *treatment* and a *dose* column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain *start_dose* and *end_dose* columns. Drop the intermediate *dose* column.

___Code___

In [34]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [35]:
treatments_clean = treatments_clean.melt(id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], \
                                         var_name='treatment', \
                                         value_name='dose')

In [36]:
treatments_clean.sample(25)
treatments_clean.shape

(700, 7)